fvm/errors/errors.go

```diff
@@ -97,3 +99,162 @@ func HandleRuntimeError(err error) error {
 	// All other errors are non-fatal Cadence errors.
 	return NewCadenceRuntimeError(runErr)
 }
+
+type prependMsgError struct {
+	msg string
+	err error
+}
+
+func (err prependMsgError) Error() string {
+	return fmt.Sprintf("%s: %v", err.msg, err.err)
+}
+
+func (err prependMsgError) Unwrap() error {
+	return err.err
+}
+
+// TODO(patrick): combine ErrorCodeChecker, Error and Failure interface.
+// We can also combine CodedError and CodedFailure once the interface is
+// unified.
+type ErrorCodeChecker interface {
+	// TODO(patrick): add Code() ErrorCode
+	// TODO(patrick): add IsFailure() bool
+
+	HasErrorCode(code ErrorCode) bool
+
+	error
+}
+
+func HasErrorCode(err error, code ErrorCode) bool {
+	var checker ErrorCodeChecker
+	return As(err, &checker) && checker.HasErrorCode(code)
+}
+
+type codedError struct {
+	code ErrorCode
+
+	// The first error in the list is the root cause.
+	errors []error
```

Comment:
Thanks for the idea. 

@janezpodhostnik  and I actually discussed a few improvement ideas for error handling and they are summarized here : https://github.com/dapperlabs/flow-go/issues/6364

For the error code, we do have a solution that is to provide a separate method `FirstCode` which will go through the chain of wrapped errors, and see if they are an instance of Error, and return the code of the first instance of Error, which is most likely the root cause.

Reply:
Rather than FirstCode, we should probably return the highest priority error code (sorted by failure vs error, then use first code as tie breaker)

Reply:
I'm not sure the use case for sorting errors by code or priority. Is it to distinguish failure and error?

If we hit a failure, then it's a fatal error, we might just turn all wrapped errors into an error string, and include it in a fmt.Errorf error. The fatal error will be used to crash the component. In other words, when we hit a failure, the failure error won't be wrapped into a sentinel error. 

This way, the caller will either get a failure or a chain of wrapped non-fatal errors (sentinel errors). 

With a chain of wrapped non-fatal errors, a caller still have full access of all the errors, although they can't be accessed by index (like `errors[len(errors) - 1]`), any error is still accessible by unwrapping.

Reply:
If we get rid of failure type, I think its better that the root cause is just the first coded error (in the error.Unwrap stack). I don't think any sorting is needed. But we definitely need to still present the entire stack so its clear which way to read it i.e:

- [Code 123] this happened
- then this happened: [Code 23] asfasf
- then some non-coded error
- which caused the transaction to fail.